integration-libs/opf/cta/core/services/opf-dynamic-cta.service.ts

```diff
@@ -0,0 +1,254 @@
+      product: this.currentProductService.getProduct(),
+      additionalData: this.fillAdditionalData(),
+    }).pipe(
+      take(1),
+      map(({ product, additionalData }) => {
+        return {
+          ctaProductItems: [{ productId: product?.code, quantity: 1 }],
+          paymentAccountIds: paymentAccountIds,
+          scriptLocations: [scriptLocation],
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  protected fillAdditionalData(): Observable<Array<KeyValuePair>> {
+    return combineLatest({
+      language: this.languageService.getActive(),
+      currency: this.currencyService.getActive(),
+    }).pipe(
+      take(1),
+      map(({ language, currency }) => {
+        console.log('language', language);
+        return [
+          {
+            key: 'locale',
+            value: 'en-US',
+          },
+          {
+            key: 'currency',
+            value: currency,
+          },
+          {
+            key: 'scriptIdentifier',
+            value: this.getNewScriptIdentifier(),
+          },
+        ];
+      })
+    );
+  }
+
+  initiateEvents() {
+    if (!this.isOnsiteMessagingInit) {
+      this.registerScriptReadyEvent();
+      this.listenScriptReadyEvent();
+
+      this.isCartPage && this.cartChangedListener();
+      !this.isCartPage && this.productChangedListener();
+
+      this.isOnsiteMessagingInit = true;
+    }
+  }
+
+  stopEvents() {
+    if (this.isOnsiteMessagingInit) {
+      this.subList.forEach((sub) => sub.unsubscribe());
+      this.subList = [];
+      this.globalFunctionsFacade.removeGlobalFunctions(
+        GlobalFunctionsDomain.GLOBAL
+      );
+      this.isOnsiteMessagingInit = false;
+    }
+  }
+
+  protected getNewScriptIdentifier(): string {
+    this.scriptIdentifiers.push(String(this.scriptIdentifiers.length + 1));
+    return String(
+      this.scriptIdentifiers[this.scriptIdentifiers.length - 1]
+    ).padStart(4, '0');
+  }
+
+  protected registerScriptReadyEvent() {
+    this.globalFunctionsFacade.registerGlobalFunctions({
+      paymentSessionId: '',
+      domain: GlobalFunctionsDomain.GLOBAL,
+    });
+  }
+
+  protected listenScriptReadyEvent() {
+    const sub = this.opfCtaFacade.listenScriptReadyEvent().subscribe({
+      next: (id: string) => {
+        this.isCartPage && this.dispatchCartEvents([id]);
+        !this.isCartPage && this.dispatchProductEvents([id]);
+      },
+    });
+    this.subList.push(sub);
+  }
+
+  protected dispatchCartEvents(scriptIdentifiers: Array<string>): void {
+    this.activeCartFacade
+      .takeActive()
+      .pipe(
+        take(1),
+        map((cart: Cart) => {
+          return { total: cart.totalPrice?.value };
+        }),
+        tap((cartTotalPrice) => {
+          const window = this.winRef.nativeWindow as any;
+          const dispatchEvent = window?.dispatchEvent;
+          if (dispatchEvent) {
+            dispatchEvent(
+              new CustomEvent('cartChanged', {
```

Comment:
Can we move `cartChanged` event name to the some const variable called `OPF_CART_CHANGED_EVENT_NAME` in case we need to re-use it somewhere else?

Reply:
oki, shouldn't be a prob as long as server script is aligned with it.